### PR TITLE
rebase--interactive: rewrite write_basic_state() and init_basic_state() in C

### PR DESCRIPTION
**Pull request by agrn**
This rewrites ``write_basic_state()`` and ``init_basic_state()`` in C. The relevant commits are the last two.

@chriscool @dscho @stefanbeller

### CODE REVIEW

File: builtin/rebase--helper.c [hunk 1]
```diff
@@ -88,13 +92,23 @@ int cmd_rebase__helper(int argc, const char **argv, const char *prefix)
...
 			   N_("squash onto")),
 		OPT_STRING(0, "upstream", &upstream, N_("upstream"),
 			   N_("the upstream commit")),
+		OPT_STRING(0, "head-name", &head_name, N_("head-name"), N_("head name")),
+		OPT_STRING('S', "gpg-sign", &opts.gpg_sign, N_("gpg-sign"),
+			   N_("GPG-sign commits")),
+		OPT_STRING(0, "strategy", &opts.strategy, N_("strategy"),
+			   N_("rebase strategy")),
+		OPT_STRING(0, "strategy-opts", &raw_strategies, N_("strategy-opts"),
+			   N_("strategy option")),
```
Comment (chriscool) [inline builtin/rebase--helper.c]: Shouldn't this be "strategy options" (with an "s" at the end of "options"?

File: builtin/rebase--helper.c [hunk 2]
```diff
@@ -44,6 +47,25 @@ static int get_revision_ranges(const char *upstream, const char *onto,
...
 }
 
+static int init_basic_state(struct replay_opts *opts, const char *head_name,
+			    const char *onto, const char *orig_head)
+{
+	struct stat sb;
+	FILE *interactive;
+
+	if (stat(path_state_dir(), &sb) && mkdir_in_gitdir(path_state_dir()))
+		return error(_("could not create temporary %s"), path_state_dir());
```
Comment (chriscool) [inline builtin/rebase--helper.c]: Maybe using error_errno() would be better as I am not sure that mkdir_in_gitdir() would output an error message.

File: builtin/rebase--helper.c [hunk 3]
```diff
@@ -44,6 +47,25 @@ static int get_revision_ranges(const char *upstream, const char *onto,
...
+
+	delete_reflog("REBASE_HEAD");
+
+	interactive = fopen(path_interactive(), "w");
+	if (!interactive)
+		return error(_("could not mark as interactive"));
```
Comment (chriscool) [inline builtin/rebase--helper.c]: Here also maybe use error_errno().

**Comment by agrn** [comment]
I also replaced “rebase--interactive” by “rebase -i” in the commit messages.

File: sequencer.h [hunk 4]
```diff
@@ -116,3 +116,7 @@ int prepare_branch_to_be_rebased(struct replay_opts *opts, const char *commit);
 void print_commit_summary(const char *prefix, const struct object_id *oid,
 			  unsigned int flags);
 #endif
+
+void parse_strategy_opts(struct replay_opts *opts, char *raw_opts);
+int write_basic_state(struct replay_opts *opts, const char *head_name,
+		      const char *onto, const char *orig_head);
```
Comment (chriscool) [inline sequencer.h]: I wonder if write_basic_state() really needs to be public in sequencer.{h,c}. Do you have plan to reuse it?

**Comment by chriscool** [comment]
Except for my last comment, this looks good to me! Thanks!

File: sequencer.h (hunk 4 above)
Comment (agrn) [reply sequencer.h]: No, I did it to have access to ``rebase_path_*()``.

Comment (chriscool) [reply sequencer.h]: Ok, then.

Comment (agrn) [reply sequencer.h]: Perhaps it would be better to make those we need public, like I did with the todo-list file?

Comment (chriscool) [reply sequencer.h]: Yeah, that's another possible solution. Maybe you can just send everything as it is and wait for possible comments about this.

**Comment by dscho** [comment]
@agrn this PR has been integrated via b49ef560ed66, correct?

**Pull request closed without merging**